src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -69,16 +89,24 @@ public int Take()
             }
 
             var index = _index;
-            var wasLastBlock = block.Next == null;
 
             if (index < block.End)
             {
                 _index = index + 1;
                 return block.Array[index];
             }
 
+            return TakeMultiBlock();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private int TakeMultiBlock()
+        {
+            var block = _block;
+            var wasLastBlock = block.Next == null;
```

Comment:
*Nooo* 😨 . You can't move the `wasLastBlock` check after the `index < block.End` check, because you'll end up with corrupted/missing request data.

[This comment](https://github.com/aspnet/KestrelHttpServer/blob/7bd027cb6b8e13a4533651636f23076a8575755e/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs#L902) explains why. The `wasLastBlock` pattern is used in about 10 places, so it's not commented everywhere. That's also why `MemoryPoolBlock.End` is volatile. We started checking `block.Next` before comparing `index` to `block.End` when we discovered #782.


The simple fix is probably to check `wasLastBlock` where we did before and pass it as a parameter to `TakeMultiBlock`, but maybe there's a more elegant fix.